### DISCUSSION TITLE
New data written to AccessStream not being decoded

### DISCUSSION THREAD

**Discussion by bhgiblin1** [ffmpeginteropx/FFmpegInteropX#245] [Q&A]
Hi,

I am using the FFmpegInteropX library to decode H.264 frames retrieved over a socket. I write one frame to a InMemoryRandomAccessStream, then CreateFromStreamAsync and set up my MediaElement. 

However, when I read the next frame and write it to the same InMemoryRandomAccessStream, the frame is never decoded and displayed. I do not see any decoder utilization on my PC (which I do see on the first frame when initializing the library). I tried looping the video, setting the Seek position and a ton of other things - but nothing seems to be doing the trick. 

I do see this in the FFmpegInteropMSS.cpp file, 
`return AVERROR_EOF;  // Let FFmpeg know that we have reached eof` https://github.com/ffmpeginteropx/FFmpegInteropX/blob/master/FFmpegInterop/FFmpegInteropMSS.cpp#L2123
Is it not trying to read again after that? I was not sure. 

Thanks for any ideas/suggestions!

**Comment by brabebhin** [comment]
Seems like a threading issue. When you write, the stream position moves, and ffmpeg will try to read at a position that resulted after your write, so it will see no data available.

**Reply by bhgiblin1** [reply]
Even if I seek to 0 on the InMemoryRandomAccessStream after writing more frames to it - nothing happens. I would expect that it would start the video over again. I also tried setting the looping mode on the MediaPlayer, along with the PlaybackSession.Position, but those didn't change anything either. 
Is there something I need to do to 'kick' FFmpegInteropX into decoding again? Just for arguments sake if I do manually create the FFmpegInteropMSS again, it decodes all of the frames in the stream.

Interestingly enough, even if I set the seek position to 0 before CreateFromStreamAsync, it never changes from 0.

**Reply by brabebhin** [reply]
This is likely because the ffmpeginterop has quite an elaborate internal state, so immutability is probably the safest approach in your case.

What kind of socket are you using? A StreamSocket from UWP? I think you will need some advanced handling of the stream data to trick ffmpeginterop into thinking it is a continuous stream, and also prevent it from having different positions. You need to somehow write to the input stream without affecting its position.

Technically, I think ffmpeginterop doesn't need the input stream to be a IRandomAccessStream. It could be an IInputStream (like the one the socket returns) but it wouldn't have seeking capability. We do not currently support this, but I think it should be easy to hack it from the existing code.

**Reply by lukasf** [reply]
When the Stream is passed to ffmpeg on first start of playback, ffmpeg will get the stream size. So after reading the one frame, it knows it is at the end of stream and will not continue playback. Even if you'd somehow trick ffmpeg into continued playback, it would also notice that the position has been changed to 0, and will seek back to the expected position of the next frame (after the first frame). This approach cannot work.

The only way to get this running, like @brabebhin said, is to implement a stream interface. On every read call, you need to provide the requested data, pushing out frame after frame. We can tell ffmpeg that a stream cannot be seeked, so theoretically you could also implement IRandomAccessStream. This would require little changes on our side. You'd just have to implement the read method and return proper values for all properties, all other methods can return E_NOTSUPPORTED. IInputStream would be easier to implement (it does not have any write methods), and it is the logical choice since it is by definition a read-only non-seekable stream. I think we could add support for that, but it would first require some work from our side. To test your approach, I'd recommend you start with a custom IRandomAccessStream.

What language are you using for implementing the socket and stream handling?

**Reply by bhgiblin1** [reply]
Thanks for the detailed replies. 
I am using C# along with a StreamSocket - reading from the InputStream and writing the frames to the IRandomAccessStream. I do have a barebones C++ application working that has the StreamSocket implemented (I was considering the use of FFmpegInteropX.FFmpegUWP, implementing the ffmpeg calls myself. However, the complexity of the MediaSampleProvider implementation deterred me from that approach). 

@lukasf you mentioned that the approach of tricking ffmpeg into continued playback would not work. However, the scenario you mentioned actually sounds like it could solve some of my appending to the stream seeking issues. If I write a new frame and the seek position changes, if ffmpeg reverts back to where the previous frame ended - that sounds like a good thing, no? 

I will start with your advice of implementing a custom IRandomAccessStream for now - thank you!

**Reply by lukasf** [reply]
What I ment is that ffmpeg will seek the stream to where it is expected. Say after decode of frame 1, you are at position X. You seek back to 0 and put frame 2 there. But ffmpeg will now see that position is modified and seek back to X, falsely expecting that the next frame is at X.

Additionally, it is not really possible for you to hook into the decoding loop. FFmpeg will usually even decode multiple frames ahead using multiple threads. So when frame 1 is decoded and returned, ffmpeg might have already read and started decoding of frame 2 and 3 (which are not even in the stream). There are multiple things that will break that approach. Custom stream is the only solution.

If performance is very important, I'd recommend you to write the complete stream and data handling pipeline in native code (C++/CX or C++/winrt). Otherwise, the GC might kick in just when data is requested, which could result in lags during playback. Not saying that this must be a problem, but it could be, depending on app, hardware and memory pressure.

**Comment by brabebhin** [comment]
I used to play round with this at some point. I think implementing IRandomAccessStream and using an InMemoryRandomAccessStream is the way to go. You need to keep in mind 2 positions, one for writing, and one for reading. The reading position will be the position you expose in the interface. When a read request comes into ffmpeg, make sure the position in the backing stream is the read position, before reading data from the stream. When writing stuff into the stream, seek to the write position (which should be well ahead of the read position, ideally) and do your write.

Just remember to make everything thread-safe.

**Reply by bhgiblin1** [reply]
Got it, thanks. 
I'll work on that implementation, and hopefully avoid needing to make changes to the main FFmpegInterop library. 

I do have control of the data on the other side as well, and would like to have some plan B in my head. 
If I wrote the H.264 frames to a RTMP stream (or another protocol), in real time - would I be able to just provide FFmpegInterop with the URI and it handle the rest?

**Reply by lukasf** [reply]
Alternatively, you could try to just read and hold the last IBuffer(s) from the socket, and return data directly from the IBuffer (to avoid copying from and to memory stream). In a different project, I have a native C++/CX BufferRandomAccessStream that takes an IBuffer and implements IRandomAccessStream over that buffer without copying memory. It is a different use case of course, but if you are interested in a native IRandomAccessStream implementation for reference purposes, I could share it.

**Reply by lukasf** [reply]
Sure, if you do have full control over the stream, then going for a real streaming protocol would probably be the best solution. I do not use streaming in my own apps, but others have reported RTMP to work fine. You can use the CreateFromUriAsync methods for that. Only the "secure" protocols (RTMPS) do not work yet due to missing openssl, but we are working on that.
